pkg/gohai/network/network.go

```diff
@@ -6,29 +6,292 @@
 // Package network regroups collecting information about the network interfaces
 package network
 
-func getNetworkInfo() (networkInfo map[string]interface{}, err error) {
-	networkInfo = make(map[string]interface{})
+import (
+	"errors"
+	"fmt"
+	"net"
 
+	"github.com/DataDog/datadog-agent/pkg/gohai/utils"
+)
+
+// ErrAddressNotFound means no such address could be found
+var ErrAddressNotFound = errors.New("address not found")
+
+// Interface holds information about a specific interface
+type Interface struct {
+	// Name is the name of the interface
+	Name string `json:"name"`
+	// IPv4Network is the ipv4 address for the host
+	IPv4Network utils.Value[string] `json:"ipv4-network"`
+	// IPv6Network is the ipv6 address for the host
+	IPv6Network utils.Value[string] `json:"ipv6-network"`
+	// MacAddress is the mac address for the host
+	MacAddress utils.Value[string] `json:"macaddress"`
+	// IPv4 is the list of IPv4 addresses for the interface
+	IPv4 []string `json:"ipv4"`
+	// IPv4 is the list of IPv6 addresses for the interface
+	IPv6 []string `json:"ipv6"`
+}
+
+// Info holds network metadata about the host
+type Info struct {
+	// Interfaces is the list of interfaces which are up
+	Interfaces []Interface `json:"interfaces"`
+	// MacAddress is a mac address of the host
+	MacAddress string `json:"macaddress"`
+	// IPAddress is an IPv4 address of the host
+	IPAddress string `json:"ipaddress"`
+	// IPAddressV6 is an IPv6 address of the host
+	IPAddressV6 utils.Value[string] `json:"ipaddressv6"`
+}
+
+// CollectInfo collects the network information.
+func CollectInfo() (*Info, error) {
+	info := &Info{}
+	err := fillNetworkInfo(info)
+	if err != nil {
+		return info, err
+	}
+
+	interfaces, err := getMultiNetworkInfo()
+	if err == nil && len(interfaces) > 0 {
+		info.Interfaces = interfaces
+	}
+	return info, err
+}
+
+func ifacesToJSON(ifaces []Interface) (interface{}, []string) {
+	ret := make([]interface{}, len(ifaces))
+	warnings := []string{}
+
+	for idx, iface := range ifaces {
+		ifaceJSON := map[string]interface{}{
+			"name": iface.Name,
+			"ipv4": iface.IPv4,
+			"ipv6": iface.IPv6,
+		}
+
+		values := map[string]utils.Value[string]{
+			"ipv4-network": iface.IPv4Network,
+			"ipv6-network": iface.IPv6Network,
+			"macaddress":   iface.MacAddress,
+		}
```

Comment:
Why don't we use `json.Marshal` here ?

Reply:
When a `Value` contains an error it shouldn't be in the resulting JSON.
This is how it worked before the whole new API and that's a behavior I tried to keep.

There is no clean workaround, even redefining how something of type `Value` is marshaled wouldn't work since there is no way to dynamically say whether it should be included in the output (at best I could have a `null` but it's still not really the same as not being in the output JSON)

Reply:
Even with the `omitempty` label ?

Reply:
Yes because `Value{value: nil, err: <some error>, initialized: true}` is not a default value, so it would be included.
There is no way to specify what is an "empty" value for a type (there were proposals but they were refused), except just reimplementing the whole JSON marshaling function